File: ledger-models-rust/fintekkers/wrappers/models/floating_rate_note.rs
//! Typed wrapper around an FRN-shaped `SecurityProto`.
//!
//! FRNs carry shared bond fields on `bond_details` plus FRN-specific extras
//! (spread over the reference rate, reference rate index, reset frequency)
//! on `frn_extension`.

use chrono::{Datelike, NaiveDate};
use rust_decimal::Decimal;

use crate::fintekkers::models::security::index::IndexTypeProto;
use crate::fintekkers::models::security::{
    BondDetailsProto, CouponFrequencyProto, CouponTypeProto, FrnExtensionProto,
    ProductTypeProto, SecurityProto,
};
use crate::fintekkers::models::util::{DecimalValueProto, LocalDateProto};
use crate::fintekkers::wrappers::models::utils::errors::Error;
use crate::fintekkers::wrappers::models::utils::serialization::ProtoSerializationUtil;

pub struct FloatingRateNote {
    pub proto: SecurityProto,
}

impl FloatingRateNote {
    pub fn from_proto(proto: SecurityProto) -> Result<Self, Error> {
        let pt = ProductTypeProto::from_i32(proto.product_type)
            .unwrap_or(ProductTypeProto::ProductTypeUnknown);
        if pt == ProductTypeProto::TreasuryFrn {
            Ok(FloatingRateNote { proto })
        } else {
            Err(Error::NotABondSecurity)
        }
    }

    pub fn spread(&self) -> Option<Decimal> {
        self.proto
            .frn_extension
            .as_ref()
            .and_then(|f| f.spread.as_ref())
            .and_then(|d| ProtoSerializationUtil::deserialize_decimal(d).ok())
    }

    pub fn reference_rate_index(&self) -> IndexTypeProto {
        self.proto
            .frn_extension
            .as_ref()
            .and_then(|f| IndexTypeProto::from_i32(f.reference_rate_index))
            .unwrap_or(IndexTypeProto::UnknownIndexType)
    }

    pub fn reset_frequency(&self) -> CouponFrequencyProto {
        self.proto
            .frn_extension
            .as_ref()
            .and_then(|f| CouponFrequencyProto::from_i32(f.reset_frequency))
            .unwrap_or(CouponFrequencyProto::UnknownCouponFrequency)
    }

    pub fn from_pricer_inputs(
        face_value: Decimal,
        coupon_rate: Decimal,
        coupon_type: CouponTypeProto,
        coupon_frequency: CouponFrequencyProto,
        issue_date: NaiveDate,
        maturity_date: NaiveDate,
        spread: Decimal,
        reference_rate_index: IndexTypeProto,
        reset_frequency: CouponFrequencyProto,
    ) -> SecurityProto {
        SecurityProto {
            product_type: ProductTypeProto::TreasuryFrn as i32,
            bond_details: Some(BondDetailsProto {
                coupon_rate: Some(decimal_proto(coupon_rate)),
                coupon_type: coupon_type as i32,
                coupon_frequency: coupon_frequency as i32,
                face_value: Some(decimal_proto(face_value)),
                issue_date: Some(date_proto(issue_date)),
                dated_date: None,
                maturity_date: Some(date_proto(maturity_date)),
                issuance_info: vec![],
            }),
            frn_extension: Some(FrnExtensionProto {
                spread: Some(decimal_proto(spread)),
                reference_rate_index: reference_rate_index as i32,
                reset_frequency: reset_frequency as i32,
            }),
            ..Default::default()
        }
    }
}

fn decimal_proto(d: Decimal) -> DecimalValueProto {
    DecimalValueProto {
        arbitrary_precision_value: d.to_string(),
    }
}

fn date_proto(d: NaiveDate) -> LocalDateProto {
    LocalDateProto {
        year: d.year() as u32,
        month: d.month(),
        day: d.day(),
    }
}

File: ledger-models-python/fintekkers/wrappers/models/security/floating_rate_note.py
"""FloatingRateNote wrapper.

Specialises BondSecurity for Treasury Floating Rate Notes. Shared bond
fields live on SecurityProto.bond_details; FRN-specific extras (spread,
reference rate index, reset frequency) live on SecurityProto.frn_extension.
"""

from datetime import date
from decimal import Decimal
from typing import Optional

from fintekkers.models.security.security_pb2 import (
    SecurityProto,
    BondDetailsProto,
    FrnExtensionProto,
)
from fintekkers.models.security.product_type_pb2 import ProductTypeProto
from fintekkers.models.security.index.index_type_pb2 import IndexTypeProto
from fintekkers.models.security.coupon_frequency_pb2 import CouponFrequencyProto
from fintekkers.wrappers.models.security.bond_security import (
    BondSecurity,
    _decimal_value,
    _local_date,
)


class FloatingRateNote(BondSecurity):
    """Wraps a SecurityProto with product_type=TREASURY_FRN."""

    def get_spread(self) -> Optional[Decimal]:
        self._ensure_hydrated("spread")
        if not self.proto.HasField("frn_extension"):
            return None
        ext = self.proto.frn_extension
        if not ext.HasField("spread"):
            return None
        value = ext.spread.arbitrary_precision_value
        if value == "":
            return None
        return Decimal(value)

    def get_reference_rate_index(self) -> IndexTypeProto:
        self._ensure_hydrated("reference_rate_index")
        if not self.proto.HasField("frn_extension"):
            return IndexTypeProto.UNKNOWN_INDEX_TYPE
        return self.proto.frn_extension.reference_rate_index

    def get_reset_frequency(self) -> CouponFrequencyProto:
        self._ensure_hydrated("reset_frequency")
        if not self.proto.HasField("frn_extension"):
            return CouponFrequencyProto.UNKNOWN_COUPON_FREQUENCY
        return self.proto.frn_extension.reset_frequency

    @staticmethod
    def from_pricer_inputs(
        face_value: Decimal,
        coupon_rate: Decimal,
        coupon_type: int,
        coupon_frequency: int,
        issue_date: date,
        maturity_date: date,
        spread: Decimal,
        reference_rate_index: int,
        reset_frequency: int,
    ) -> SecurityProto:
        """Builds an FRN SecurityProto with both bond_details and
        frn_extension populated, and product_type=TREASURY_FRN.
        """
        bond_details = BondDetailsProto(
            coupon_rate=_decimal_value(coupon_rate),
            coupon_type=coupon_type,
            coupon_frequency=coupon_frequency,
            face_value=_decimal_value(face_value),
            issue_date=_local_date(issue_date),
            maturity_date=_local_date(maturity_date),
        )
        frn_extension = FrnExtensionProto(
            spread=_decimal_value(spread),
            reference_rate_index=reference_rate_index,
            reset_frequency=reset_frequency,
        )
        return SecurityProto(
            product_type=ProductTypeProto.TREASURY_FRN,
            bond_details=bond_details,
            frn_extension=frn_extension,
        )
